File: algorithms/algorithms-cplusplus/coding_interviews_2/Question_0066_ConstuctArray.cc
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> multiply(const vector<int>& A) {
    if (A.size() <= 1) {
      return vector<int>();
    }
    vector<int> res(A.size());
    int tmp = 1, n = A.size();
    for (int i = 0; i < n; i++) {
      res[i] = tmp;
      tmp *= A[i];
    }
    tmp = 1;
    for (int i = n - 1; i >= 0; i--) {
      res[i] *= tmp;
      tmp *= A[i];
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/coding_interviews_2/Question_0066_ConstuctArray.java
package com.lanceolata.coding.algorithms.java.coding_interviews_2;

public class Question_0066_ConstuctArray {
    public int[] multiply(int[] A) {
        if (A == null || A.length == 0) {
            return null;
        }
        int[] res = new int[A.length];
        res[0] = 1;
        int tmp = 1;
        for (int i = 1; i < A.length; i++) {
            tmp *= A[i - 1];
            res[i] = tmp;
        }
        tmp = 1;
        for (int i = A.length - 2; i >= 0; i--) {
            tmp *= A[i + 1];
            res[i] *= tmp;
        }
        return res;
    }
}
